configprofiles: TestDataDriven failed

Epic: CRDB-18499

configprofiles.TestDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11116118?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11116118?buildTab=artifacts#/) on master @ [f295bd861a3a427652b19c2254d2401ebb4a3c8e](https://github.com/cockroachdb/cockroach/commits/f295bd861a3a427652b19c2254d2401ebb4a3c8e):


```
=== RUN   TestDataDriven
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/f07519a9a253a7694f6e67cf7554fe40/logTestDataDriven252320693
    test_log_scope.go:81: use -show-logs to present logs inline
=== CONT  TestDataDriven
    datadriven_test.go:132: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f07519a9a253a7694f6e67cf7554fe40/logTestDataDriven252320693
--- FAIL: TestDataDriven (209.06s)
=== RUN   TestDataDriven/virtual-app-repl
    datadriven_test.go:84: condition failed to evaluate within 45s: expected 6 tasks to be completed, got 0
    --- FAIL: TestDataDriven/virtual-app-repl (62.77s)
=== RUN   TestDataDriven/replication-source
    datadriven_test.go:84: condition failed to evaluate within 45s: expected 6 tasks to be completed, got 0
    --- FAIL: TestDataDriven/replication-source (62.54s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/multi-tenant @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30225

configprofiles.TestDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11169264?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11169264?buildTab=artifacts#/) on master @ [4fe2a80d81c6fc5a3da3c7c44c5fc38da67e0367](https://github.com/cockroachdb/cockroach/commits/4fe2a80d81c6fc5a3da3c7c44c5fc38da67e0367):


```
=== RUN   TestDataDriven
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/12957e36f12eb56cf35cd3a5dad40bd1/logTestDataDriven4273186083
    test_log_scope.go:81: use -show-logs to present logs inline
=== CONT  TestDataDriven
    datadriven_test.go:132: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/12957e36f12eb56cf35cd3a5dad40bd1/logTestDataDriven4273186083
--- FAIL: TestDataDriven (29.83s)
=== RUN   TestDataDriven/virtual-app
    datadriven_test.go:89: error scanning '&{0x4006875320 <nil>}': pq: service unavailable for target tenant (application)
    --- FAIL: TestDataDriven/virtual-app (1.72s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

configprofiles.TestDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11185624?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11185624?buildTab=artifacts#/) on master @ [886515ffc2f1485f47bc3448db93abe3f0217f03](https://github.com/cockroachdb/cockroach/commits/886515ffc2f1485f47bc3448db93abe3f0217f03):


```
=== RUN   TestDataDriven
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/12957e36f12eb56cf35cd3a5dad40bd1/logTestDataDriven3452377934
    test_log_scope.go:81: use -show-logs to present logs inline
=== CONT  TestDataDriven
    datadriven_test.go:132: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/12957e36f12eb56cf35cd3a5dad40bd1/logTestDataDriven3452377934
--- FAIL: TestDataDriven (27.80s)
=== RUN   TestDataDriven/virtual-app
    datadriven_test.go:89: error scanning '&{0x4009bd10e0 <nil>}': pq: service unavailable for target tenant (application)
    --- FAIL: TestDataDriven/virtual-app (1.82s)
=== RUN   TestDataDriven/virtual-app-repl
    datadriven_test.go:89: error scanning '&{0x400b4f4b40 <nil>}': pq: service unavailable for target tenant (application)
    --- FAIL: TestDataDriven/virtual-app-repl (2.09s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

configprofiles.TestDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11187844?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11187844?buildTab=artifacts#/) on master @ [d87a04bf528cf3d8f242ceb539451cd14aa7a3cf](https://github.com/cockroachdb/cockroach/commits/d87a04bf528cf3d8f242ceb539451cd14aa7a3cf):


```
=== RUN   TestDataDriven
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/f07519a9a253a7694f6e67cf7554fe40/logTestDataDriven2278900871
    test_log_scope.go:81: use -show-logs to present logs inline
=== CONT  TestDataDriven
    datadriven_test.go:132: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f07519a9a253a7694f6e67cf7554fe40/logTestDataDriven2278900871
--- FAIL: TestDataDriven (312.85s)
=== RUN   TestDataDriven/virtual-app
    datadriven_test.go:84: condition failed to evaluate within 3m45s: expected 5 tasks to be completed, got 0
    --- FAIL: TestDataDriven/virtual-app (237.87s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

configprofiles.TestDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11195800?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11195800?buildTab=artifacts#/) on master @ [7a597b66ac4c4f63a9a3c115b2d41c20b1393816](https://github.com/cockroachdb/cockroach/commits/7a597b66ac4c4f63a9a3c115b2d41c20b1393816):


```
=== RUN   TestDataDriven
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/12957e36f12eb56cf35cd3a5dad40bd1/logTestDataDriven2749586949
    test_log_scope.go:81: use -show-logs to present logs inline
=== CONT  TestDataDriven
    datadriven_test.go:132: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/12957e36f12eb56cf35cd3a5dad40bd1/logTestDataDriven2749586949
--- FAIL: TestDataDriven (58.93s)
=== RUN   TestDataDriven/replication-source
    datadriven_test.go:89: error scanning '&{0x4004d23b00 <nil>}': pq: service unavailable for target tenant (application)
    --- FAIL: TestDataDriven/replication-source (1.39s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

configprofiles.TestDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11222498?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11222498?buildTab=artifacts#/) on master @ [cdf6d1574474d96466b1481048ace4485c2f24c9](https://github.com/cockroachdb/cockroach/commits/cdf6d1574474d96466b1481048ace4485c2f24c9):


```
=== RUN   TestDataDriven
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/12957e36f12eb56cf35cd3a5dad40bd1/logTestDataDriven4003944995
    test_log_scope.go:81: use -show-logs to present logs inline
=== CONT  TestDataDriven
    datadriven_test.go:132: -- test log scope end --
--- FAIL: TestDataDriven (60.23s)
=== RUN   TestDataDriven/virtual-app-repl
    datadriven_test.go:89: error scanning '&{0x4002ce1200 <nil>}': pq: service unavailable for target tenant (application)
    --- FAIL: TestDataDriven/virtual-app-repl (1.48s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

configprofiles.TestDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11226725?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11226725?buildTab=artifacts#/) on master @ [782a30119cd1854ca6211e5d343ad8f156f7d0c1](https://github.com/cockroachdb/cockroach/commits/782a30119cd1854ca6211e5d343ad8f156f7d0c1):


```
=== RUN   TestDataDriven
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/12957e36f12eb56cf35cd3a5dad40bd1/logTestDataDriven3776976689
    test_log_scope.go:81: use -show-logs to present logs inline
=== CONT  TestDataDriven
    datadriven_test.go:132: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/12957e36f12eb56cf35cd3a5dad40bd1/logTestDataDriven3776976689
--- FAIL: TestDataDriven (25.64s)
=== RUN   TestDataDriven/virtual-app
    datadriven_test.go:89: error scanning '&{0x4009663440 <nil>}': pq: service unavailable for target tenant (application)
    --- FAIL: TestDataDriven/virtual-app (1.82s)
=== RUN   TestDataDriven/virtual-app-repl
    datadriven_test.go:89: error scanning '&{0x4004bc7200 <nil>}': pq: service unavailable for target tenant (application)
    --- FAIL: TestDataDriven/virtual-app-repl (1.51s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

configprofiles.TestDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11256806?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11256806?buildTab=artifacts#/) on master @ [b91c919df727958cb68056bdb382b89b878aad83](https://github.com/cockroachdb/cockroach/commits/b91c919df727958cb68056bdb382b89b878aad83):


```
=== RUN   TestDataDriven
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/12957e36f12eb56cf35cd3a5dad40bd1/logTestDataDriven1894688850
    test_log_scope.go:81: use -show-logs to present logs inline
=== CONT  TestDataDriven
    datadriven_test.go:132: -- test log scope end --
--- FAIL: TestDataDriven (31.54s)
=== RUN   TestDataDriven/replication-source
    datadriven_test.go:89: error scanning '&{0x4006b74000 <nil>}': pq: service unavailable for target tenant (application)
    --- FAIL: TestDataDriven/replication-source (1.46s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/server/BUILD.bazel
+++ b/pkg/server/BUILD.bazel
@@ -491,6 +491,7 @@ go_test(
         "//pkg/kv/kvserver/kvserverpb",
         "//pkg/kv/kvserver/kvstorage",
         "//pkg/kv/kvserver/liveness/livenesspb",
+        "//pkg/multitenant",
         "//pkg/roachpb",
         "//pkg/rpc",
         "//pkg/security",

--- a/pkg/server/server_controller_test.go
+++ b/pkg/server/server_controller_test.go
@@ -15,10 +15,14 @@ import (
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/multitenant"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
@@ -81,6 +85,7 @@ func TestSQLErrorUponInvalidTenant(t *testing.T) {
 	require.NoError(t, err)
 
 	err = db.Ping()
+	require.NotNil(t, err)
 	require.Regexp(t, `service unavailable for target tenant \(nonexistent\)`, err.Error())
 }
 
@@ -108,4 +113,58 @@ func TestSharedProcessServerInheritsTempStorageLimit(t *testing.T) {
 
 	tss := ts.(*testTenant)
 	require.Equal(t, int64(specialSize), tss.SQLCfg.TempStorageConfig.Mon.Limit())
+}
+
+// TestServerSQLConn checks that the SQLConn() method on the
+// SystemLayer() of TestServerInterface works even when non-specific
+// SQL connection requests are redirected to a secondary tenant.
+func TestServerSQLConn(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	s := serverutils.StartServerOnly(t, base.TestServerArgs{
+		DefaultTestTenant: base.TestControlsTenantsExplicitly,
+	})
+	defer s.Stopper().Stop(ctx)
+	systemTenant := s.SystemLayer()
+
+	// Start some secondary tenant servers.
+	secondaryTenantExtNoName, err := s.StartTenant(ctx, base.TestTenantArgs{
+		TenantID: roachpb.MustMakeTenantID(2),
+	})
+	require.NoError(t, err)
+
+	secondaryTenantExtNamed, err := s.StartTenant(ctx, base.TestTenantArgs{
+		TenantName: "hello",
+		TenantID:   roachpb.MustMakeTenantID(10),
+	})
+	require.NoError(t, err)
+
+	secondaryTenantSh, _, err := s.StartSharedProcessTenant(ctx, base.TestSharedProcessTenantArgs{
+		TenantName: "world",
+	})
+	require.NoError(t, err)
+	multitenant.DefaultTenantSelect.Override(ctx, &systemTenant.ClusterSettings().SV, "world")
+
+	for _, tc := range []struct {
+		testName     string
+		tbName       string
+		sqlInterface serverutils.ApplicationLayerInterface
+	}{
+		{"system", "foo", systemTenant},
+		{"secondary-external-noname", "bar", secondaryTenantExtNoName},
+		{"secondary-external-named", "baz", secondaryTenantExtNamed},
+		{"secondary-shared", "qux", secondaryTenantSh},
+	} {
+		t.Run(tc.testName, func(t *testing.T) {
+			_, err = tc.sqlInterface.InternalExecutor().(isql.Executor).
+				Exec(ctx, "create-table", nil, "CREATE TABLE defaultdb."+tc.tbName+" (i INT)")
+			require.NoError(t, err)
+
+			conn := tc.sqlInterface.SQLConn(t, "defaultdb")
+			var unused int
+			assert.NoError(t, conn.QueryRowContext(ctx, "SELECT count(*) FROM "+tc.tbName).Scan(&unused))
+		})
+	}
 }

--- a/pkg/server/testserver.go
+++ b/pkg/server/testserver.go
@@ -54,6 +54,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/deprecatedshowranges"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire"
 	"github.com/cockroachdb/cockroach/pkg/sql/physicalplan"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/catconstants"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
 	"github.com/cockroachdb/cockroach/pkg/storage"
@@ -495,7 +496,9 @@ func (ts *testServer) SQLConnE(dbName string) (*gosql.DB, error) {
 
 // SQLConnForUserE is part of the serverutils.ApplicationLayerInterface.
 func (ts *testServer) SQLConnForUserE(userName string, dbName string) (*gosql.DB, error) {
-	return openTestSQLConn(userName, dbName, ts.Stopper(),
+	return openTestSQLConn(
+		userName, dbName, catconstants.SystemTenantName,
+		ts.Stopper(),
 		ts.topLevelServer.loopbackPgL,
 		ts.cfg.SQLAdvertiseAddr,
 		ts.cfg.Insecure,
@@ -787,7 +790,15 @@ func (t *testTenant) SQLConnE(dbName string) (*gosql.DB, error) {
 
 // SQLConnForUserE is part of the serverutils.ApplicationLayerInterface.
 func (t *testTenant) SQLConnForUserE(userName string, dbName string) (*gosql.DB, error) {
-	return openTestSQLConn(userName, dbName, t.Stopper(),
+	tenantName := t.t.tenantName
+	if !t.Cfg.DisableSQLListener {
+		// This tenant server has its own SQL listener. It will not accept
+		// a "cluster" connection parameter.
+		tenantName = ""
+	}
+	return openTestSQLConn(
+		userName, dbName, tenantName,
+		t.Stopper(),
 		t.pgL,
 		t.Cfg.SQLAdvertiseAddr,
 		t.Cfg.Insecure,

--- a/pkg/server/testserver_sqlconn.go
+++ b/pkg/server/testserver_sqlconn.go
@@ -13,10 +13,13 @@ package server
 import (
 	"context"
 	gosql "database/sql"
+	"fmt"
 	"net"
 	"net/url"
+	"strings"
 	"time"
 
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/netutil"
@@ -43,6 +46,7 @@ const useLoopbackListener = false
 // of serverutils.ApplicationLayerInterface.
 func openTestSQLConn(
 	userName, dbName string,
+	tenantName roachpb.TenantName,
 	stopper *stop.Stopper,
 	// When useLoopbackListener is set, only this is used:
 	pgL *netutil.LoopbackListener,
@@ -53,13 +57,21 @@ func openTestSQLConn(
 	cleanupFn := func() {}
 	var goDB *gosql.DB
 
+	opts := url.Values{}
+	if tenantName != "" && !strings.HasPrefix(dbName, "cluster:") {
+		opts.Add("options", fmt.Sprintf("-ccluster=%s", tenantName))
+	}
+	if insecure || useLoopbackListener {
+		opts.Add("sslmode", "disable")
+	}
+
 	if useLoopbackListener {
 		pgurl := url.URL{
 			Scheme:   "postgres",
 			User:     url.User(userName),
 			Host:     "unused",
 			Path:     dbName,
-			RawQuery: "sslmode=disable",
+			RawQuery: opts.Encode(),
 		}
 		// TODO(sql): consider using pgx for tests instead of lib/pq.
 		connector, err := pq.NewConnector(pgurl.String())
@@ -76,9 +88,16 @@ func openTestSQLConn(
 			return nil, err
 		}
 		pgURL.Path = dbName
-		if insecure {
-			pgURL.RawQuery = "sslmode=disable"
+
+		// Add the common query options decided above to those prepared by
+		// PGUrlE().
+		qv := pgURL.Query()
+		for k, v := range opts {
+			qv[k] = v
 		}
+		pgURL.RawQuery = qv.Encode()
+
+		// Open the connection.
 		goDB, err = gosql.Open("postgres", pgURL.String())
 		if err != nil {
 			cleanupFn()

